arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,561 @@
+inline unsigned int
+integer_log2( unsigned int value )
+{
+  unsigned int l = 0;
+  while( ( value >> l ) > 1 )
+  {
+    ++l;
+  }
+  return l;
+}
+
+// ----------------------------------------------------------------------------
+// Intersect a region with some image boundaries
+template < typename PixType >
+void
+check_region_boundaries( vgl_box_2d< int >& bbox,
+                         vil_image_view< PixType > const& img )
+{
+  vgl_box_2d< int > boundaries{ 0, static_cast< int >( img.ni() ),
+                                0, static_cast< int >( img.nj() ) };
+  bbox = vgl_intersection( boundaries, bbox );
+}
+
+// ----------------------------------------------------------------------------
+// Point an image view to a rectangular region in the image
+template < typename PixType, typename BoxType >
+void
+point_view_to_region( vil_image_view< PixType > const& src,
+                      vgl_box_2d< BoxType > const& region,
+                      vil_image_view< PixType >& dst )
+{
+  // Early exit case, no crop required
+  if( region.min_x() == 0 && region.min_y() == 0 &&
+      region.max_x() == static_cast< BoxType >( src.ni() ) &&
+      region.max_y() == static_cast< BoxType >( src.nj() ) )
+  {
+    dst = src;
+    return;
+  }
+
+  // Validate boundaries
+  vgl_box_2d< BoxType > to_crop = region;
+  check_region_boundaries( to_crop, src );
+
+  // Make sure width and height are non-zero
+  if( to_crop.width() == 0 || to_crop.height() == 0 )
+  {
+    return;
+  }
+
+  // Perform crop
+  dst = vil_crop( src, to_crop.min_x(), to_crop.width(), to_crop.min_y(),
+                  to_crop.height() );
+}
+
+// ----------------------------------------------------------------------------
+// Alternative call for the above
+template < typename PixType, typename BoxType >
+vil_image_view< PixType >
+point_view_to_region( vil_image_view< PixType > const& src,
+                      vgl_box_2d< BoxType > const& region )
+{
+  vil_image_view< PixType > output;
+  point_view_to_region( src, region, output );
+  return output;
+}
+
+// ----------------------------------------------------------------------------
+// An optimized (unsafe if used incorrectly) function which populates
+// a n^p dimensional histogram from integer image 'input' given the
+// resolution of each channel of the histogram, and a bitshift value
+// which maps each value to its channel step for the histogram
+template < class InputType >
+void
+populate_image_histogram(
+  vil_image_view< InputType > const& input, unsigned* hist_top_left,
+  unsigned const bitshift, std::ptrdiff_t const* hist_steps )
+{
+  // Get image properties
+  auto const ni = input.ni();
+  auto const nj = input.nj();
+  auto const np = input.nplanes();
+  auto const istep = input.istep();
+  auto const jstep = input.jstep();
+  auto const pstep = input.planestep();
+
+  // Filter image
+  InputType const* row = input.top_left_ptr();
+  for( decltype( +nj ) j{ 0 }; j < nj; ++j, row += jstep )
```

Comment:
It seems that this `decltype( +nj ) j{ 0 }` hack doesn't work to remove `const` for Visual Studio.  You could try:
`auto j = decltype(nj){0};`.

Personally, I think `decltype` also makes the code harder to read.  I don't see a strong argument for this over explicitly using `unsigned j = 0`, but I'm okay with `decltype` as long as it works reliably.

Reply:
@mwoehlke-kitware was advocating for it, so perhaps he has an argument for it.

However, I'm inclined to agree with you, especially in this context. Most times I've written `decltype` across these related PRs, it should evaluate to `unsigned`. From a maintenance perspective, I see the appeal if the argument to `decltype` is likely to be refactored, but in this case it's primarily `vil_image_views` which I doubt will change their API or be replaced with another type in the near future.

Reply:
I know @mwoehlke-kitware really likes `auto` / `decltype`, and there are places where it really comes in handy.  However, I would argue that this is not really one of them.  `vil` has been doing loops over image pixels with `unsigned` type for 20+ years.  That type is not going to change and you could argue that `unsigned j=0` makes this loop better follow VXL style.  There are probably hundreds of examples in VXL of `for( unsigned j=0; j < nj; ++j, row += jstep )`.